peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+                self = cls.object_cache[foo + 1]
+
+            else:
+                # this instance is not
+                self = super().__new__(cls)
+
+            # assignment to an object which has been initialized before,
+            # breaking the invariant a read-only attribute can be assigned to
+            # only during its initialization?
+            self.foo = foo
+
+            cls.object_cache[foo] = self
+            return self
+
+To my understanding, properly detecting this problem would require type checkers
+to keep track of the "level of initialization" of an object.
+
+This issue doesn't seem to impact ``__init__``, since it's rather uncommon to
+ever rebind ``self`` within it to any other object, and type checkers could
+flag the action as whole.
+
+
+Extending initialization
+------------------------
+
+Mechanisms such as :func:`dataclasses.__post_init__` or attrs' `initialization hooks <https://www.attrs.org/en/stable/init.html#hooking-yourself-into-initialization>`_
+augment initialization by providing a set of dunder hooks which will be called
+once during instance creation. The current rules would disallow assignment in those
+hooks. Specifying any single method in the pep isn't enough, as the naming and
+functionality differs between mechanisms (``__post_init__`` vs ``__attrs_post_init__``).
+
+``ReadOnly[ClassVar[...]]`` and ``__init_subclass__``
+-----------------------------------------------------
+
+Should this be allowed?
+
+.. code-block:: python
+
+    class URI:
+        protocol: ReadOnly[ClassVar[str]] = ""
+
+        def __init_subclass__(cls, protocol: str = "") -> None:
+            cls.foo = protocol
+
+    class File(URI, protocol="file"): ...
+
+``Final`` in protocols
+----------------------
+
+It's been `suggested <https://discuss.python.org/t/expanding-readonly-to-normal-classes-protocols/67359/45>`_
+to clarify in this PEP whether ``Final`` should be supported by protocols.
+
+
+Footnotes
+=========
+
+.. [#overriding_property]
+    Pyright in strict mode disallows non-property overrides.
+    Mypy does not impose this restriction and allows an override with a plain attribute.
+    `[Pyright playground] <https://pyright-play.net/?strict=true&code=MYGwhgzhAEBiD28BcBYAUNT0D6A7ArgLYBGApgE5LQCWuALuultACakBmO2t1d22ACgikQ7ADTQCJClVp0AlNAC0APmgA5eLlKoMzLMNEA6PETLloAXklmKjPZgACAB3LxnFOgE8mWNpylzIRF2RVUael19LHJSOnxyXGhDdhNAuzR7UEgYACEwcgEEeHkorHTKCIY0IA>`_
+    `[mypy playground] <https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=6f860a865c5d13cce07d6cbb08b9fb85>`_
+
+.. [#runtime]
+    This PEP focuses solely on the type-checking behavior. Nevertheless, it should
+    be desirable the name is read-only at runtime.
+
+.. [#invalid_typevar]
+    The implied type variable is not valid in this context; it has been used for
+    the ease of demonstration. See `ClassVar <https://typing.readthedocs.io/en/latest/spec/class-compat.html#classvar>`_.
```

Comment:
What are we seeing `ClassVar` for here?

Reply:
There's this section in the linked spec:
> Note that a ClassVar parameter cannot include any type variables, regardless of the level of nesting: ClassVar[T] and ClassVar[list[set[T]]] are both invalid if T is a type variable.

Reply:
Ah, I see. To be honest, I think this whole footnote can be deleted. There is nothing in the section that it is footnoting that defines `T` as specifically a typevar, requiring this disclaimer; T is just a stand-in name for "some arbitrary type name". This is very typical in typing documentation/spec and doesn't require further comment, IMO.